aiu_fms_testing_utils/utils/__init__.py

```diff
@@ -180,61 +252,134 @@ def __sample_requests(
         prompt_length_max (int): filters out prompts larger than this value.
         enforce_sizes (List[int]): sample request will grab a prompt with this length if available.
-        enforce_heterogeneous (bool): Pads all prompts within batch size to nearest multiple of 64.
+        enforce_heterogeneous (bool): Pads all prompts within batch size to nearest multiple of `pad_multiple`.
+            However, if enforce_sizes is not empty, it will set enforce_heteogeneous to False.
         pad_multiple (int): Used only when enforce_heterogeneous is True or enforce_sizes is not empty, asserts that prompt_length would be padded to this multiple
         List[Tuple[str, int]]: a filtered dataset
+        truncation (bool): If true will truncate to an enforced size if the size does not exist. Only to be used with enforce_sizes, otherwise
+        will be ignored
+        _cached_dataset_key (optional[str]): The key to the dataset if enabling caching of encoded datasets
+
+    Returns:
+        List[Tuple[str, int]]
     """
 
+    assert prompt_length_max >= prompt_length_min, (
+        "Please enter valid prompt length max/min values"
+    )
+
+    if enforce_sizes is None:
+        enforce_sizes = []
+
+    if enforce_heterogeneous and enforce_sizes:
+        warnings.warn(
+            f"{enforce_heterogeneous=} and {enforce_sizes=}, these two are not designed to be used at the same time. Forcing enforce_heterogeneous to False"
+        )
+        enforce_heterogeneous = False
+
     # Based on min/max prompt length, one can back out the number of possible heterogeneous values
     max_heterogeneous_combinations = (prompt_length_max // pad_multiple) - (
         (prompt_length_min - 1) // pad_multiple
     )
 
     # Filter out sequences that are too long or too short
+    dataset: List[Tuple[str, int]] = []
     filtered_dataset: List[Tuple[str, int]] = []
     enforced_dataset: List[Tuple[str, int]] = []
 
     # To track sizes seen
     seen_sizes: List[int] = []
 
+    sample_size_counter: dict[int, int] = {}
+    # first int is the size to truncate to, second int is size of text to grab from
+    enforce_sizes_with_truncation: List[Tuple[int, int]] = []
+
+    if truncation and not enforce_sizes:
+        warnings.warn(
+            f"truncation and enforce_sizes should be used together, whereas {truncation=} and {enforce_sizes=}, hence no truncation will happen",
+            stacklevel=2,
+        )
+
+    if (
+        _cached_dataset_key is not None
+        and _cached_dataset_key in __cached_encoded_datasets
+    ):
+        dataset = __cached_encoded_datasets[_cached_dataset_key]
+    else:
+        # Loop to check create filtered dataset
+        for i in range(len(prompt_list)):
+            # Tokenize the prompts and completions.
+            prompt = prompt_list[i]
+            prompt_token_ids = tokenizer.encode(prompt, return_tensors="pt").squeeze(0)
+
+            prompt_len = len(prompt_token_ids)
```

Comment:
this can be parallelized by using the APIs more efficiently passing in the whole dataset as a single call to encode